libs/entitlement/tools/securitymanager-scanner/src/main/java/org/elasticsearch/entitlement/tools/securitymanager/scanner/SecurityCheckClassVisitor.java

```diff
@@ -0,0 +1,262 @@
+            if (opcode == GETSTATIC && descriptor.endsWith("Permission;")) {
+                var permissionType = Type.getType(descriptor);
+                if (permissionType.getSort() == Type.ARRAY) {
+                    permissionType = permissionType.getElementType();
+                }
+                try {
+                    var clazz = Class.forName(permissionType.getClassName());
+                    if (Permission.class.isAssignableFrom(clazz)) {
+                        this.permissionType = permissionType.getInternalName();
+                    }
+                } catch (ClassNotFoundException e) {
+                    handleException(permissionType.getClassName(), e);
+                }
+
+                var objectType = Type.getObjectType(owner);
+                try {
+                    var clazz = Class.forName(objectType.getClassName());
+                    Arrays.stream(clazz.getDeclaredFields())
+                        .filter(f -> Modifier.isStatic(f.getModifiers()) && Modifier.isFinal(f.getModifiers()))
+                        .filter(f -> f.getName().equals(name))
+                        .findFirst()
+                        .ifPresent(x -> {
+                            if (Permission.class.isAssignableFrom(x.getType())) {
+                                try {
+                                    x.setAccessible(true);
+                                    var p = (Permission) (x.get(null));
+                                    this.runtimePermissionType = p.getName();
+                                } catch (IllegalAccessException | InaccessibleObjectException e) {
+                                    handleException(x.getName(), e);
+                                }
+                            }
+                        });
+
+                } catch (ClassNotFoundException | NoClassDefFoundError | UnsatisfiedLinkError e) {
+                    handleException(objectType.getClassName(), e);
+                }
+            }
+        }
+
+        @Override
+        public void visitLdcInsn(Object value) {
+            super.visitLdcInsn(value);
+            if (permissionType != null && permissionType.equals("java/lang/RuntimePermission")) {
+                this.runtimePermissionType = value.toString();
+            }
+        }
+
+        @Override
+        public void visitMethodInsn(int opcode, String owner, String name, String descriptor, boolean isInterface) {
+            super.visitMethodInsn(opcode, owner, name, descriptor, isInterface);
+            if (opcode == INVOKEVIRTUAL
+                || opcode == INVOKESPECIAL
+                || opcode == INVOKESTATIC
+                || opcode == INVOKEINTERFACE
+                || opcode == INVOKEDYNAMIC) {
+
+                if (SECURITY_MANAGER_INTERNAL_NAME.equals(owner)) {
+                    boolean callerIsPublic = (methodAccess & ACC_PUBLIC) != 0
+                        && (classAccess & ACC_PUBLIC) != 0
```

Comment:
Hmm. Could there be public methods on non-public classes that are still part of the API surface area? I'm thinking the API could export an interface, and then the implementation of that interface could be a non-public class that contains a `SecurityManager` check.

Reply:
I'll change this to an enum to include the "public method on non-public class" case